### PR TITLE
Changed internal representation of messages and added LLM required ordering of messages.

### PR DESCRIPTION
**Pull request by NanditaRao**
Changed the Message Handlers to use message factory to create new message from a user. Added a parent message pointer to each message in a thread. Represented messages so that the order is always alternating between user message and bot message.

### CODE REVIEW

File: src/main/java/com/meta/cp4m/message/Message.java [hunk 1]
```diff
@@ -32,6 +36,8 @@ static Identifier threadId(Identifier id1, Identifier id2) {
 
   Role role();
```
Comment (NanditaRao) [inline src/main/java/com/meta/cp4m/message/Message.java:37]: Add priority order to enum and use that when ordering messages in ThreadState.

File: src/main/java/com/meta/cp4m/llm/HuggingFaceLlamaPlugin.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ public T handle(ThreadState<T> threadState) throws IOException {
     Optional<String> prompt = promptCreator.createPrompt(threadState);
     if (prompt.isEmpty()) {
       return threadState.newMessageFromBot(
-          Instant.now(), "I'm sorry but that request was too long for me.");
+          Instant.now(), "I'm sorry but that request was too long for me.", threadState.tail());
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/llm/HuggingFaceLlamaPlugin.java:53]: isn't the last message always the tail? seems like `ThreadState` should manage that logic

File: src/main/java/com/meta/cp4m/message/MessageFactory.java [hunk 3]
```diff
@@ -46,7 +46,8 @@ T newMessage(
       Identifier senderId,
       Identifier recipientId,
       Identifier instanceId,
-      Role role);
+      Role role,
+      Message parentMessage);
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/MessageFactory.java:50]: Looks like `parentMessage` is always null, we should just eliminate it and you can use `withParent` if needed

File: src/main/java/com/meta/cp4m/message/Message.java [hunk 4]
```diff
@@ -32,13 +37,22 @@ static Identifier threadId(Identifier id1, Identifier id2) {
...
 
+  Message parentMessage();
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/Message.java:40]:
```suggestion
  @Nullable Message parentMessage();
```

File: src/main/java/com/meta/cp4m/message/ThreadState.java [hunk 5]
```diff
@@ -33,20 +33,17 @@ private ThreadState(T message) {
...
     messageFactory = old.messageFactory;
     Preconditions.checkArgument(
         old.tail().threadId().equals(newMessage.threadId()),
         "all messages in a thread must have the same thread id");
     List<T> messages = old.messages;
-    if (newMessage.timestamp().isBefore(old.tail().timestamp())) {
-      this.messages =
-          Stream.concat(messages.stream(), Stream.of(newMessage))
-              .sorted(Comparator.comparing(Message::timestamp))
+    T mWithParentMessage = newMessage.role() == Role.USER ? (T) newMessage.addParentMessage(old.tail()): newMessage;
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/ThreadState.java:42]: You have an unchecked cast here, this should very rarely need to happen. Basically what this message is saying is that there is that the compiler cannot guarantee that `newMessage` and `old.tail()` will be of the same type, (which is the whole reason for using an strongly typed language 🙂). You are in a bit of a pickle here though as to get around this you'd really need a new generic on Message indicating the type of the parent message which isn't ideal. Let me think on this for a bit

File: src/main/java/com/meta/cp4m/message/Message.java [hunk 6]
```diff
@@ -32,13 +37,22 @@ static Identifier threadId(Identifier id1, Identifier id2) {
...
 
   enum Role {
-    ASSISTANT,
-    USER,
-    SYSTEM
+    ASSISTANT(0),
+    USER(1),
+    SYSTEM(2);
+
+    public final Integer priority;
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/Message.java:51]: So go ahead and get a copy of "Effective Java" by Joshua Bloch, it's a really great book for the foundation and sets a lot of Java conventions. 

"Item 16: In public classes, use accessor methods, not public fields"
"Item 61: Prefer primitive types to boxed primitives"
```suggestion
    private final int priority;
    
    public int priority() {
      return priority;
    }
```

File: src/main/java/com/meta/cp4m/message/Message.java [hunk 7]
```diff
@@ -13,13 +13,17 @@
 
 public interface Message {
 
-  static Identifier threadId(Identifier id1, Identifier id2) {
+  public static Identifier threadId(Identifier id1, Identifier id2) {
     if (id1.compareTo(id2) <= 0) {
       return Identifier.from(id1.toString() + '|' + id2);
     }
     return Identifier.from(id2.toString() + '|' + id1);
   }
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/Message.java:20]: can we make this package-private again now?

File: src/main/java/com/meta/cp4m/llm/HuggingFaceLlamaPlugin.java (hunk 2 above)
Comment (NanditaRao) [reply src/main/java/com/meta/cp4m/llm/HuggingFaceLlamaPlugin.java:53]: Addressed this!

File: src/main/java/com/meta/cp4m/message/Message.java (hunk 4 above)
Comment (NanditaRao) [reply src/main/java/com/meta/cp4m/message/Message.java:40]: Addressed this!

File: src/main/java/com/meta/cp4m/message/Message.java (hunk 6 above)
Comment (NanditaRao) [reply src/main/java/com/meta/cp4m/message/Message.java:51]: Addressed this!

File: src/main/java/com/meta/cp4m/message/Message.java (hunk 7 above)
Comment (NanditaRao) [reply src/main/java/com/meta/cp4m/message/Message.java:20]: Addressed this!

File: src/main/java/com/meta/cp4m/message/Message.java [hunk 8]
```diff
@@ -32,13 +36,26 @@ static Identifier threadId(Identifier id1, Identifier id2) {
...
+    USER(1),
+    SYSTEM(2);
+
+    private final int priority;
+
+    private Role(Integer priority){
+      this.priority = priority;
+    }
+
+    public int getPriority(){
```
Comment (hunterjackson) [inline src/main/java/com/meta/cp4m/message/Message.java:56]:
```suggestion
    public int priority(){
```

I never told you but I've been using a fluent api style so lets stick with that

**Pull request closed without merging**